File: app/src/main/java/com/manimarank/spell4wiki/activities/CommonWebActivity.java
package com.manimarank.spell4wiki.activities;

import android.content.Intent;
import android.os.Bundle;
import android.text.SpannableString;
import android.text.style.ForegroundColorSpan;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import com.manimarank.spell4wiki.R;
import com.manimarank.spell4wiki.activities.base.BaseActivity;
import com.manimarank.spell4wiki.fragments.LanguageSelectionFragment;
import com.manimarank.spell4wiki.fragments.WebViewFragment;
import com.manimarank.spell4wiki.listerners.OnLanguageSelectionListener;
import com.manimarank.spell4wiki.utils.PrefManager;
import com.manimarank.spell4wiki.utils.constants.AppConstants;
import com.manimarank.spell4wiki.utils.constants.EnumTypeDef.ListMode;

public class CommonWebActivity extends BaseActivity {

    private boolean isWiktionaryWord = false;
    private WebViewFragment fragment = new WebViewFragment();
    private String languageCode = "";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_common_web_view);

        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }


        PrefManager pref = new PrefManager(getApplicationContext());
        languageCode = pref.getLanguageCodeWiktionary();

        Bundle bundle = getIntent().getExtras();
        if (bundle != null) {

            String title = "";
            if (bundle.containsKey(AppConstants.TITLE)) {
                title = bundle.getString(AppConstants.TITLE);
                setTitle(title);
            }

            if (bundle.containsKey(AppConstants.IS_WIKTIONARY_WORD))
                isWiktionaryWord = bundle.getBoolean(AppConstants.IS_WIKTIONARY_WORD);

            if (bundle.containsKey(AppConstants.LANGUAGE_CODE))
                languageCode = bundle.getString(AppConstants.LANGUAGE_CODE);

            loadFragment(fragment);
        }

    }

    private void loadFragment(Fragment fragment) {
        FragmentManager fm = getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fm.beginTransaction();
        fragmentTransaction.replace(R.id.frameLayout, fragment);
        fragmentTransaction.commit();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.web_view_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_favorite:
                return true;
            case R.id.action_share:
                fragment.shareLink();
                return true;
            case R.id.action_refresh:
                fragment.refreshWebPage();
                return true;
            case R.id.action_forward:
                fragment.forwardWebPage();
                return true;
            case R.id.action_backward:
                fragment.backwardWebPage();
                return true;
            case R.id.action_open_in_browser:
                fragment.openInAppBrowser();
                return true;
            case R.id.action_copy_link:
                fragment.copyLink();
                return true;
            case android.R.id.home:
                finish();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void setupLanguageSelectorMenuItem(Menu menu) {
        menu.findItem(R.id.menu_lang_selector).setVisible(isWiktionaryWord);
        if (isWiktionaryWord) {
            MenuItem item = menu.findItem(R.id.menu_lang_selector);
            View rootView = item.getActionView();
            TextView selectedLang = rootView.findViewById(R.id.txtSelectedLanguage);
            selectedLang.setText(this.languageCode.toUpperCase());
            rootView.setOnClickListener(v -> {
                loadLanguages();
            });
        }
    }

    private void loadLanguages() {
        if (isWiktionaryWord) {
            OnLanguageSelectionListener callback = langCode -> {
                if (!languageCode.equals(langCode)) {
                    this.languageCode = langCode;
                    invalidateOptionsMenu();
                    if (fragment != null)
                        fragment.loadWordWithOtherLang(langCode);
                }
            };
            LanguageSelectionFragment languageSelectionFragment = new LanguageSelectionFragment(this);
            languageSelectionFragment.init(callback, ListMode.TEMP, languageCode);
            languageSelectionFragment.show(getSupportFragmentManager());
        }
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        boolean result = super.onPrepareOptionsMenu(menu);
        changeMenuButtonStyle(menu.findItem(R.id.action_forward), fragment.canGoForward());
        changeMenuButtonStyle(menu.findItem(R.id.action_backward), fragment.canGoBackward());
        setupLanguageSelectorMenuItem(menu);
        return result;
    }

    private void changeMenuButtonStyle(MenuItem menuItem, boolean isAllow) {
        if (menuItem != null) {
            SpannableString s = new SpannableString(menuItem.getTitle());
            s.setSpan(new ForegroundColorSpan(ContextCompat.getColor(getApplicationContext(), isAllow ? R.color.black : R.color.light_gray)), 0, s.length(), 0);
            menuItem.setEnabled(isAllow);
            menuItem.setTitle(s);
        }
    }


    public void updateList(String word) {
        if (fragment != null)
            fragment.hideRecordButton(word);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == AppConstants.RC_UPLOAD_DIALOG) {
            if (data != null && data.hasExtra(AppConstants.WORD)) {
                updateList(data.getStringExtra(AppConstants.WORD));
            }
        }
    }

}

File: app/src/main/java/com/manimarank/spell4wiki/ui/webui/CommonWebActivity.kt
package com.manimarank.spell4wiki.ui.webui

import android.content.Intent
import android.os.Bundle
import android.text.SpannableString
import android.text.style.ForegroundColorSpan
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import androidx.appcompat.widget.Toolbar
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import com.manimarank.spell4wiki.R
import com.manimarank.spell4wiki.data.db.DBHelper
import com.manimarank.spell4wiki.data.db.dao.WikiLangDao
import com.manimarank.spell4wiki.data.prefs.PrefManager
import com.manimarank.spell4wiki.ui.common.BaseActivity
import com.manimarank.spell4wiki.ui.languageselector.LanguageSelectionFragment
import com.manimarank.spell4wiki.ui.listerners.OnLanguageSelectionListener
import com.manimarank.spell4wiki.utils.GeneralUtils
import com.manimarank.spell4wiki.utils.constants.AppConstants
import com.manimarank.spell4wiki.utils.constants.ListMode
import java.util.Locale

class CommonWebActivity : BaseActivity() {
    private var wikiLangDao: WikiLangDao? = null
    private var isWiktionaryWord = false
    private val fragment: WebViewFragment = WebViewFragment()
    private var languageCode: String? = ""
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_common_web_view)

        val pref = PrefManager(applicationContext)
        languageCode = pref.languageCodeSpell4WikiAll
        wikiLangDao = DBHelper.getInstance(applicationContext).appDatabase.wikiLangDao

        // Title & Sub title
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        val bundle = intent.extras
        if (bundle != null) {
            val title: String?
            if (bundle.containsKey(AppConstants.TITLE)) {
                title = bundle.getString(AppConstants.TITLE)
                setTitle(title)
            }
            if (bundle.containsKey(AppConstants.IS_WIKTIONARY_WORD))
                isWiktionaryWord = bundle.getBoolean(AppConstants.IS_WIKTIONARY_WORD)
            if (bundle.containsKey(AppConstants.LANGUAGE_CODE)) {
                languageCode = bundle.getString(AppConstants.LANGUAGE_CODE)
                supportActionBar?.subtitle = GeneralUtils.getLanguageInfo(applicationContext, wikiLangDao?.getWikiLanguageWithCode(languageCode))
            }
            loadFragment(fragment)
        }
    }

    private fun loadFragment(fragment: Fragment?) {
        val fm = supportFragmentManager
        val fragmentTransaction = fm.beginTransaction()
        fragmentTransaction.replace(R.id.frameLayout, fragment!!)
        fragmentTransaction.commit()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.web_view_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_favorite -> true
            R.id.action_share -> {
                fragment.shareLink()
                true
            }
            R.id.action_refresh -> {
                fragment.refreshWebPage()
                true
            }
            R.id.action_forward -> {
                fragment.forwardWebPage()
                true
            }
            R.id.action_backward -> {
                fragment.backwardWebPage()
                true
            }
            R.id.action_open_in_browser -> {
                fragment.openInAppBrowser()
                true
            }
            R.id.action_copy_link -> {
                fragment.copyLink()
                true
            }
            android.R.id.home -> {
                finish()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun setupLanguageSelectorMenuItem(menu: Menu) {
        menu.findItem(R.id.menu_lang_selector).isVisible = isWiktionaryWord
        if (isWiktionaryWord) {
            val item = menu.findItem(R.id.menu_lang_selector)
            val rootView = item.actionView ?: return
            val selectedLang = rootView.findViewById<TextView>(R.id.txtSelectedLanguage)
            selectedLang.text = languageCode?.toUpperCase(Locale.ROOT) ?: ""
            rootView.setOnClickListener { loadLanguages() }
        }
    }

    private fun loadLanguages() {
        if (isWiktionaryWord) {
            val callback = object : OnLanguageSelectionListener {
                override fun onCallBackListener(langCode: String?) {
                    if (languageCode != langCode) {
                        languageCode = langCode
                        supportActionBar?.subtitle = GeneralUtils.getLanguageInfo(applicationContext, wikiLangDao?.getWikiLanguageWithCode(langCode))
                        invalidateOptionsMenu()
                        fragment.loadWordWithOtherLang(langCode)
                    }
                }
            }
            val languageSelectionFragment = LanguageSelectionFragment(this)
            languageSelectionFragment.init(callback, ListMode.TEMP, languageCode)
            languageSelectionFragment.show(supportFragmentManager)
        }
    }

    override fun onPrepareOptionsMenu(menu: Menu): Boolean {
        val result = super.onPrepareOptionsMenu(menu)
        changeMenuButtonStyle(menu.findItem(R.id.action_forward), fragment.canGoForward())
        changeMenuButtonStyle(menu.findItem(R.id.action_backward), fragment.canGoBackward())
        setupLanguageSelectorMenuItem(menu)
        return result
    }

    private fun changeMenuButtonStyle(menuItem: MenuItem?, isAllow: Boolean) {
        if (menuItem != null) {
            val s = SpannableString(menuItem.title)
            s.setSpan(ForegroundColorSpan(ContextCompat.getColor(applicationContext, if (isAllow) R.color.black else R.color.light_gray)), 0, s.length, 0)
            menuItem.isEnabled = isAllow
            menuItem.title = s
        }
    }

    fun updateList(word: String?) {
        fragment.hideRecordButton(word!!)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == AppConstants.RC_UPLOAD_DIALOG) {
            if (data != null && data.hasExtra(AppConstants.WORD)) {
                updateList(data.getStringExtra(AppConstants.WORD))
            }
        }
    }
}